### PR TITLE
[#1501] Add unit tests for FeedbackRankQuestionDetails and FeedbackTextResponseDetails classes

### PR DESCRIPTION
**Pull request by daongochieu2810**
Fixes #1501

### CODE REVIEW

**Comment by daongochieu2810** [comment]
Hi my PR is failing the E2E tests. How do I fix this ? Thanks

File: src/test/java/teammates/common/datatransfer/questions/FeedbackRankQuestionDetailsTest.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    @Test
+    public void testValidateMinOptionsToBeRanked_sameValidValues_shouldReturnTrue() {
+        int testValue = 100;
+        FeedbackRankQuestionDetails feedbackRankQuestionDetails = new FeedbackRankOptionsQuestionDetails();
+        feedbackRankQuestionDetails.setMinOptionsToBeRanked(testValue);
+        assertEquals(feedbackRankQuestionDetails.minOptionsToBeRanked, testValue);
+    }
+
+    @Test
+    public void testValidateMinOptionsToBeRanked_differentValidValues_shouldReturnTrue() {
```
Comment (rrtheonlyone) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackRankQuestionDetailsTest.java:20]: This test is not really doing anything right? We already test that a valid value is set in the above test. (same for the other one below)

File: src/test/java/teammates/common/datatransfer/questions/FeedbackRankQuestionDetailsTest.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package teammates.common.datatransfer.questions;
+
+import org.testng.annotations.Test;
+
+import teammates.test.BaseTestCase;
+
+/**
+ * SUT: {@link FeedbackRankQuestionDetails}.
+ */
+public class FeedbackRankQuestionDetailsTest extends BaseTestCase {
```
Comment (rrtheonlyone) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackRankQuestionDetailsTest.java:10]: add a test for setting/getting duplicates too + you could also test the the default value is `Integer.MIN_VALUE`

File: src/test/java/teammates/common/datatransfer/questions/FeedbackTextResponseDetailsTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+public class FeedbackTextResponseDetailsTest extends BaseTestCase {
+    @Test
+    public void testValidateResponseAnswer_sameValues_shouldReturnTrue() {
+        String testAnswer = "Hello World";
+        FeedbackTextResponseDetails feedbackTextResponseDetails = new FeedbackTextResponseDetails(testAnswer);
+        assertEquals(feedbackTextResponseDetails.getAnswer(), "Hello World");
+    }
+
+    @Test
+    public void testValidateResponseAnswer_differentValues_shouldReturnFalse() {
```
Comment (rrtheonlyone) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackTextResponseDetailsTest.java:20]: again this isnt doing much

File: src/test/java/teammates/common/datatransfer/questions/FeedbackTextResponseDetailsTest.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        String testAnswer = "Hello World";
+        FeedbackTextResponseDetails feedbackTextResponseDetails = new FeedbackTextResponseDetails(testAnswer);
+        assertNotEquals(feedbackTextResponseDetails.getAnswer(), "World Hello");
+    }
+
+    @Test(expectedExceptions = NullPointerException.class)
+    public void testValidateAttributesOfAnswer_nullValue_shouldThrowException() {
```
Comment (rrtheonlyone) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackTextResponseDetailsTest.java:27]: you can assert is null here instead of forcing an NPE

File: src/test/java/teammates/common/datatransfer/questions/FeedbackTextResponseDetailsTest.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        FeedbackTextResponseDetails feedbackTextResponseDetails = new FeedbackTextResponseDetails(null);
+        feedbackTextResponseDetails.getAnswer().length();
+    }
+
+    @Test
+    public void testValidateAttributesOfAnswer_validValue_shouldThrowException() {
```
Comment (rrtheonlyone) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackTextResponseDetailsTest.java:33]: I think this can be removed - seems like a duplicate of earlier test

File: src/test/java/teammates/common/datatransfer/questions/FeedbackTextResponseDetailsTest.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+import org.testng.annotations.Test;
+
+import teammates.common.util.SanitizationHelper;
+import teammates.test.BaseTestCase;
+
+/**
+ * SUT: {@link FeedbackTextResponseDetails}.
+ */
+public class FeedbackTextResponseDetailsTest extends BaseTestCase {
```
Comment (rrtheonlyone) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackTextResponseDetailsTest.java:11]: I think we can test `setAnswer` and also the default constructor (that sets answer to empty string)

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Looks good! Just some minor comments

File: src/test/java/teammates/common/datatransfer/questions/FeedbackRankQuestionDetailsTest.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    public void testValidateDuplicatesAllowed_validValues_shouldReturnTrue() {
+        FeedbackRankQuestionDetails feedbackRankQuestionDetails = new FeedbackRankOptionsQuestionDetails();
+        feedbackRankQuestionDetails.setAreDuplicatesAllowed(true);
+        assertEquals(feedbackRankQuestionDetails.areDuplicatesAllowed, true);
+
+        feedbackRankQuestionDetails.setAreDuplicatesAllowed(false);
+        assertEquals(feedbackRankQuestionDetails.areDuplicatesAllowed, false);
+
+        feedbackRankQuestionDetails.setAreDuplicatesAllowed(true);
+        assertNotEquals(feedbackRankQuestionDetails.areDuplicatesAllowed, false);
```
Comment (ChooJeremy) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackRankQuestionDetailsTest.java:46]: These checks aren't necessary, they make the code more complex (uncertainty in what you're testing) and are already covered by the above equal tests.

File: src/test/java/teammates/common/datatransfer/questions/FeedbackTextResponseDetailsTest.java (hunk 4 above)
Comment (ChooJeremy) [reply src/test/java/teammates/common/datatransfer/questions/FeedbackTextResponseDetailsTest.java:20]: I think your naming of the test function should change here as well?

File: src/test/java/teammates/common/datatransfer/questions/FeedbackTextResponseDetailsTest.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        assertEquals(feedbackTextResponseDetails.getAnswerString(),
+                SanitizationHelper.sanitizeForRichText(feedbackTextResponseDetails.getAnswer()));
+    }
+
+    @Test
+    public void testValidateSetAnswer_sameValues_shouldReturnTrue() {
+        String testValue = "Hello World";
+        String emptyString = "";
+        FeedbackTextResponseDetails feedbackTextResponseDetails = new FeedbackTextResponseDetails();
+        assertEquals(feedbackTextResponseDetails.getAnswerString(),
```
Comment (ChooJeremy) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackTextResponseDetailsTest.java:45]: To be clearer for what you're testing (the function `setAnswer`), you should test that the initial value isn't `testValue` i.e. 
```suggestion
        assertNotEquals(feedbackTextResponseDetails.getAnswerString(), SanitizationHelper.sanitizeForRichText(testValue));
```

**Review by ChooJeremy — CHANGES_REQUESTED** [review]

File: src/main/java/teammates/common/datatransfer/questions/FeedbackTextResponseDetails.java [hunk 9]
```diff
@@ -3,7 +3,6 @@
 import teammates.common.util.SanitizationHelper;
 
 public class FeedbackTextResponseDetails extends FeedbackResponseDetails {
-
```
Comment (ChooJeremy) [inline src/main/java/teammates/common/datatransfer/questions/FeedbackTextResponseDetails.java:6]: This change seems unrelated?
Let's try to avoid changes in files that don't impact your PR.

File: src/test/java/teammates/common/datatransfer/questions/FeedbackRankQuestionDetailsTest.java [hunk 10]
```diff
@@ -25,28 +25,10 @@ public void testValidateMaxOptionsToBeRanked_sameValidValues_shouldReturnTrue()
...
-        feedbackRankQuestionDetails.setMaxOptionsToBeRanked(testValue);
-        assertNotEquals(feedbackRankQuestionDetails.maxOptionsToBeRanked, testDiffValue);
-    }
-
-    @Test
-    public void testValidateDuplicatesAllowed_validValues_shouldReturnTrue() {
+    public void testValidateDuplicatesAllowed_isAllowed_shouldReturnTrue() {
         FeedbackRankQuestionDetails feedbackRankQuestionDetails = new FeedbackRankOptionsQuestionDetails();
         feedbackRankQuestionDetails.setAreDuplicatesAllowed(true);
         assertEquals(feedbackRankQuestionDetails.areDuplicatesAllowed, true);
```
Comment (ChooJeremy) [inline src/test/java/teammates/common/datatransfer/questions/FeedbackRankQuestionDetailsTest.java:31]: Sorry if I wasn't clear enough. We do want to test that a change has occured, just not test an Equals and a NotEquals of a boolean in the same state.
Something along the lines of:

- Create new object
- Test that duplicates are allowed is not true
- Set duplicates allowed to true
- Test that duplicates are allowed is now true.

We can do the same for max options as well.

**Review by ChooJeremy — CHANGES_REQUESTED** [review]
Getting there! I think there might be some misunderstanding on testing the impact of the set function.

**Comment by damithc** [comment]
Looks like this is ready to review again?

**Review by rrtheonlyone — APPROVED** [review]
LGTM 👍 

Should be good to go once you fix the linting and pass the CI

**Review by rrtheonlyone — APPROVED** [review]

**Review by ChooJeremy — APPROVED** [review]
lgtm
